device-selection-explainer.md

```diff
@@ -164,6 +164,84 @@ Many platforms support various hybrid execution scenarios involving NPU, CPU, an
+
+#### **1. Operation-Type Optimization**
+
+- **Matrix multiplication (compute-bound)**:
+Use **GPUs** for large matrix operations (e.g., transformer prefill stages). GPUs achieve **~22% lower latency** and **~2× higher throughput** than NPUs for these tasks due to parallel compute units.
+    - Example: Serving Llama 70B with TensorRT-LLM on NVIDIA Hopper GPUs.
+- **Matrix-vector multiplication (memory-bound)**:
+Deploy **NPUs**, which reduce latency by **58.5%** compared to GPUs. NPUs leverage DMA for efficient memory access, ideal for LLM decode phases.
+    - Example: NPUs process TinyLlama inference **3.2× faster** than GPUs.
+- **Low-complexity operations (e.g., dot product)**:
+Assign to **CPUs**, which avoid GPU/NPU memory overhead and achieve lower latency for non-parallel tasks.
+
+#### **2. Model Architecture Considerations**
+
+- **Large Language Models (LLMs)**:
+    - **Prefill**: GPU clusters (compute-heavy)
+    - **Decode**: NPUs (memory-bound, sequential token generation)
+    - Use **disaggregated serving** to split phases across devices, boosting throughput up to **30×**.
+- **LSTM/RNN Models**:
+Prefer **GPUs**, which outperform NPUs by **2.7×** due to irregular memory access patterns.
+- **Vision Models (e.g., MobileNetV2)**:
+    - **Small batches**: NPUs (consistent latency)
+    - **Large batches**: GPUs (scaling throughput)
+
+#### **3. Batch Size and Latency Tradeoffs**
+
+| Scenario | Preferred Hardware | Rationale |
+| :-- | :-- | :-- |
+| Small batch (1-8) | NPU | 3× lower latency for video classification[^2] |
+| Large batch (>32) | GPU | Throughput scales with parallel compute[^2] |
+| Real-time SLOs | NPU + CPU | NPU for decode, CPU for lightweight ops[^3] |
+
+#### **4. Power-Constrained Deployments**
+
+- **NPUs** consume **≤50% power** of GPUs for equivalent performance, making them ideal for edge devices.
+- Use **CPU/NPU hybrids** for latency-sensitive applications requiring energy efficiency.
+
+#### **5. Dynamic Orchestration**
+
+Tools may monitor GPU/NPU utilization and automatically:
+
+- Shift decode GPUs to prefill during traffic spikes
+- Select optimal tensor parallelism strategies (e.g. separately for prefill and decode)
+- Leverage support for low-latency data movement between heterogeneous devices.
+
+By combining hardware-specific strengths with adaptive resource management, developers can achieve **2–30× throughput improvements** while maintaining strict latency targets.
+
+### Simplified guide
+
+| Factor | CPU | GPU | NPU |
+| :-- | :-- | :-- | :-- |
+| **Best For** | Sequential logic, small models | Parallel training, large batches | Edge inference, low-power AI |
+| **Power Efficiency** | Moderate | High consumption | Ultra-efficient |
+| **Latency** | 50–100 ms | 10–30 ms | 2–10 ms |
+| **Typical Use** | Preprocessing, decision trees | LLM training, computer vision | Smartphones, IoT devices |
```

Comment:
For the purposes of this explainer, perhaps this simplified guide is what we need?

I wonder if we could use qualitative terms for the latency instead of time ranges? And/or map the devices to some latency-sensitive use cases?

And perhaps we leave training out.